apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/theadpool/ThreadPoolService.java

```diff
@@ -0,0 +1,102 @@
+import org.apache.skywalking.apm.agent.core.boot.DefaultNamedThreadFactory;
+import org.apache.skywalking.apm.agent.core.logging.api.ILog;
+import org.apache.skywalking.apm.agent.core.logging.api.LogManager;
+import org.apache.skywalking.apm.agent.core.meter.MeterFactory;
+import org.apache.skywalking.apm.util.RunnableWithExceptionProtection;
+
+public class ThreadPoolService implements BootService, Runnable {
+
+    private static final ILog LOGGER = LogManager.getLogger(ThreadPoolService.class);
+
+    private volatile ScheduledFuture<?> buildMeterMetricFuture;
+
+    private static final Map<String, Function<ThreadPoolExecutor, Supplier<Double>>> METRIC_MAP = ImmutableMap.of(
+            "core_pool_size",
+            (ThreadPoolExecutor threadPoolExecutor) -> () -> {
+                return (double) threadPoolExecutor.getCorePoolSize();
+            },
+            "max_pool_size",
+            (ThreadPoolExecutor threadPoolExecutor) -> () -> (double) threadPoolExecutor.getMaximumPoolSize(),
+            "pool_size",
+            (ThreadPoolExecutor threadPoolExecutor) -> () -> (double) threadPoolExecutor.getPoolSize(),
+            "queue_size",
+            (ThreadPoolExecutor threadPoolExecutor) -> () -> (double) threadPoolExecutor.getQueue().size(),
+            "active_size",
+            (ThreadPoolExecutor threadPoolExecutor) -> () -> (double) threadPoolExecutor.getActiveCount());
+
+    @Override
+    public void prepare() throws Throwable {
+    }
+
+    @Override
+    public void boot() throws Throwable {
+        buildMeterMetricFuture = Executors.newSingleThreadScheduledExecutor(
+                        new DefaultNamedThreadFactory("ThreadPoolService-BuildMetric"))
+                .scheduleAtFixedRate(new RunnableWithExceptionProtection(
+                        this, t -> LOGGER.error("ThreadPoolService-BuildMetric failure.", t)
+                ), 0, 1, TimeUnit.SECONDS);
+
+    }
+
+    @Override
+    public void onComplete() throws Throwable {
+
+    }
+
+    @Override
+    public void shutdown() throws Throwable {
+        buildMeterMetricFuture.cancel(true);
+    }
+
+    private void buildThreadPoolMeterMetric() {
+        String threadPoolMeterName = "thread_pool";
+        String poolNameTag = "pool_name";
+        String metricTypeTag = "metric_type";
+        List<ThreadPoolSourceData> threadPoolSourceDataList = ThreadPoolRegister.getThreadPoolSourceDataList();
+        threadPoolSourceDataList.forEach(it -> {
+            if (!it.getHasBuildMetric()) {
+                String poolName = it.getThreadPoolName().name().toLowerCase();
+                ThreadPoolExecutor threadPoolExecutor = it.getThreadPoolExecutor();
+                METRIC_MAP.forEach(
+                        (key, value) -> MeterFactory.gauge(threadPoolMeterName, value.apply(threadPoolExecutor))
+                                .tag(poolNameTag, poolName).tag(metricTypeTag, key).build());
```

Comment:
You don't need to create a scheduler to create the gauge, `MeterFactory#gauge` or other `MeterFactory#metric` means you could just register once, and [MeterSender](https://github.com/apache/skywalking-java/blob/main/apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/meter/MeterSender.java) would collect the metric with scheduler.

Reply:
bacasue plugin resgister and create gauge order is not certain， so scheduler get it and create the gauge, but only create onece

Reply:
We provide you with a provider, rather than a static value. When the core is calling you, you could return the latest one. The registration is only about the name of the metric.